File: 1137-n-th-tribonacci-number/1137-n-th-tribonacci-number.py
class Solution:
    def tribonacci(self, n: int) -> int:
        
        dp = {}
        
        def result(n):
            
            if n in dp:
                return dp[n]
            
            if n==0:
                return 0

            if n==1 or n==2:
                return 1

            dp[n] = result(n-3)+result(n-3+1)+result(n-3+2)
            
            return dp[n]
        
        return result(n)
    

File: 1137-n-th-tribonacci-number/1137-n-th-tribonacci-number.java
class Solution {
    
    HashMap<Integer,Integer> dp = new HashMap<>();
    
    public int tribonacci(int n) {
        
        if(dp.containsKey(n)){
            return dp.get(n);
        }
        
        if(n==0){
            return 0;
        }    
        if(n==1||n==2){
            return 1;
        }
        dp.put(n,tribonacci(n-3)+tribonacci(n-2)+tribonacci(n-1));
        return dp.get(n);
    }
}